hercules/hercules_model.py

```diff
@@ -10,54 +10,79 @@
+    close_logging,
+    get_available_component_names,
+    get_available_component_types,
+    hercules_float_type,
+    interpolate_df,
+    load_yaml,
+    setup_logging,
+)
 
 LOGFILE = str(dt.datetime.now()).replace(":", "_").replace(" ", "_").replace(".", "_")
 
 Path("outputs").mkdir(parents=True, exist_ok=True)
 
 
-class Emulator:
-    def __init__(self, controller, hybrid_plant, h_dict, logger):
+class HerculesModel:
+    def __init__(self, input_file):
         """
-        Initializes the emulator.
+        Initializes the HerculesModel.
 
         Args:
-            controller (object): The controller object responsible for managing the simulation.
-            hybrid_plant (object): An object containing hybrid plant components.
-            h_dict (dict): A dictionary contains parameters and values for the simulation.
-            logger (object): A logger instance for logging messages during the simulation.
+            input_file (Union[str, dict]): Path to Hercules input YAML file or dictionary
+                containing input configuration.
+            controller_class (class): The controller class (not instance) to be initialized
+                with the prepared h_dict.
 
         """
 
         # Make sure output folder exists
         Path("outputs").mkdir(parents=True, exist_ok=True)
 
-        # Use the provided logger
-        self.logger = logger
+        # Set up logging
+        self.logger = self._setup_logging()
 
-        # Save the input dict to main dict
-        self.h_dict = h_dict
+        # Load and validate the input file
+        self.h_dict = self._load_hercules_input(input_file)
+
+        # Initialize the hybrid plant
+        self.hybrid_plant = HybridPlant(self.h_dict)
+
+        # Add plant component metadata to h_dict
+        self.h_dict = self.hybrid_plant.add_plant_metadata_to_h_dict(self.h_dict)
+
+        # Initialize the controller as None, to be assigned in a subsequent call
+        self._controller = None
 
         # Initialize the flattened h_dict
         self.h_dict_flat = {}
 
         # Save time step, start time and end time first
-        self.dt = h_dict["dt"]
-        self.starttime = h_dict["starttime"]  # Always 0, computed from UTC
-        self.endtime = h_dict["endtime"]  # Duration in seconds, computed from UTC
+        self.dt = self.h_dict["dt"]
+        self.starttime = self.h_dict["starttime"]  # Always 0, computed from UTC
+        self.endtime = self.h_dict["endtime"]  # Duration in seconds, computed from UTC
 
         # Save UTC timestamps
-        self.starttime_utc = h_dict["starttime_utc"]
-        self.endtime_utc = h_dict["endtime_utc"]
+        self.starttime_utc = self.h_dict["starttime_utc"]
+        self.endtime_utc = self.h_dict["endtime_utc"]
 
         # Initialize logging configuration
-        self.log_every_n = h_dict.get("log_every_n", 1)
+        self.log_every_n = self.h_dict.get("log_every_n", 1)
         self.dt_log = self.dt * self.log_every_n
```

Comment:
Unless there is a reason not to, I'd probably move this above the HybridPlant instantiation (to just after _load_hercules_input) since this stuff seems more about setting up the simulation parameters.

Reply:
Done